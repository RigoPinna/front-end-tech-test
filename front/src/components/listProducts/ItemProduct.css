.__card {
    display: inline-block;
    width: 100%;
    background: #FFF;
    border-radius: 0.625em;
    margin: 0.3125em;
    box-shadow: rgba(9, 30, 66, 0.25) 0px 1px 1px, rgba(9, 30, 66, 0.13) 0px 0px 1px 1px;
    overflow: hidden;
    position: relative;
    font-size: 1rem;
}
.__card:active {
    transform:opacity(0);
}
.__card > span{
    position: absolute;
    right: 0;
    top: 0;
    padding: 0.3125em;
    background:rgba(0, 0, 0, 0.248);
    color: #fff;
    font-size: 0.75em;
}
.__card img {
    width: 100%;
    height: 10.3125em;
    object-fit: cover;
    font-size: 1em;
}
.__card_body {
    padding:0.3125em;
    font-size: 1em;
}
.__card_body h4 {
    font-size:1.0625em;
    word-break: break-word;
    font-weight: 900;
    position: relative;


}
.__card_body > p:nth-child(2) {
    font-size: 0.75em;
    /* color:pink; */
    font-weight: 400;
    color: #929292;

}
.__card_body > p {
    color: #4e4e4e;
    font-size: 0.875em;
    word-break: break-word;
    position: relative;
}
.__card_body_description {
    
    display: -webkit-box;
    max-width: 100%;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
}
.__card_footer {
    width: 100%;
    /* height: 52px; */

    padding: 0.3125em;
    align-items: center;
    justify-content: space-between;
    font-size: 1em;

}
.__card_footer strong {
    font-size: 1.0625em;
    font-weight: 900;
}
.__card_footer button {
    margin-top:0.3125em;
    width: 100%;
    font-size: 0.875em;
}
@media screen and (min-width: 460px) {
    .__card {
        width: 300px;
        font-size: 1em;
    }
}