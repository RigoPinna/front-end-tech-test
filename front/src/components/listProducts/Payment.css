.__wrapper_payment {
    width: 100%;
    min-height: 6.25em;
    border-top: 0.0625em solid #282C34;
    padding: 0.3125em 1.25em;
    font-size: 1em;
}
.__wrapper_payment_price {
    display: flex;
    width: 100%;
    height: 3.0625em;
    justify-content: space-between;
    font-size:1em
}
.__wrapper_payment_price span{
    font-weight: normal;
    font-size:0.75em
}
@media screen and (min-width: 750px) {
    .__wrapper_payment {
        font-size:20px;
    }

}