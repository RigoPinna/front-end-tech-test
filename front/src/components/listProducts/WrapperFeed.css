.__wrapper_feed_products {
    display: flex;
    justify-content: space-evenly;
    flex-wrap: wrap;
    width: 100%;
    padding: 5px;
    margin-top: 45px;
}
.__wrapper_feed_products h2 {
    /* display: inline; */
    width: 100%;
}
/* .__wrapper_feed_products .__card {
    width: 100%;
} */

