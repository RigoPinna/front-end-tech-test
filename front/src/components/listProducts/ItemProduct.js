import React from 'react'
import PropTypes from 'prop-types'
import { useHistory } from 'react-router-dom'
import { useCart } from '../../hooks/useCart'
import { baseURL } from '../../API'

import { IconSad } from '../icons/IconSad'
import { IconAdd } from '../icons/IconAdd'

import './ItemProduct.css'
import { RatingStart } from '../ratingStarts/RatingStart'
export const ItemProduct = ( product ) => {
    const history = useHistory()
    const { addItemToCart } = useCart( product );
    return (
        <div className ='__card animate__animated animate__fadeIn' onClick={()=> history.push(`/product/${product._id}`)}>
            <span>#{product.rating}</span>
            <img src={`${baseURL}${product.image}`} alt={product.name} />
            <div className='__card_body'>
                <h4>{product.name}</h4>
                <p>{`${product.brand} - ${product.category}`}</p>
                <p className='__card_body_description'>{ product.description }</p>
                <RatingStart {...product} />
            </div>
            <div className ='__card_footer'>
                <strong>${product.price}</strong>
                <button 
                    onClick={ addItemToCart }
                    disabled={ product.countInStock < 1}>
                    {
                        (product.countInStock < 1) 
                            ? <><IconSad/>Out of stock</> 
                            : <><IconAdd />Add to item cart</>
                    }
                </button>
            </div>
        </div>
    )
}
ItemProduct.propTypes = {
    _id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    brand: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    countInStock: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
}
