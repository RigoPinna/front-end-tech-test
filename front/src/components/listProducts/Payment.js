import React from 'react'
import { useCart } from '../../hooks/useCart'

import './Payment.css'

export const Payment = () => {
    const { totalItems, totalPrice } = useCart()
    return (
        <div className="__wrapper_payment animate__animated animate__fadeIn">
            <div className='__wrapper_payment_price'>
                <strong>Bag total</strong>
                <strong><span>({ totalItems } items)</span> ${totalPrice}</strong>
            </div>
            
        </div>
    )
}
