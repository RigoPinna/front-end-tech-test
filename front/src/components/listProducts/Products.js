import React from 'react'
import { useSelector } from 'react-redux'
import { WrapperFeed } from './WrapperFeed'
import { ItemProduct } from './ItemProduct'

export const Products = () => {

    const { productsReducer } = useSelector(state => state)
    return (
        <>
            <WrapperFeed>
                {
                    (productsReducer.length > 0) 
                        ? <h2>Fount {productsReducer.length} Resoults</h2>
                        : <h2>No Productos Founts</h2>
                }
                {
                    (productsReducer.length > 0) 
                        && productsReducer.map( product => <ItemProduct key={`product-list-${product._id}`} {...product}/> )
                }
            </WrapperFeed>
        </>
    )
}
