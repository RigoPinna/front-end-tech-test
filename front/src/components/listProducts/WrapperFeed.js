import React from 'react'
import './WrapperFeed.css'
export const WrapperFeed = ({children}) => {
    return (
        <section className="__wrapper_feed_products">
            { children }
        </section>
    )
}
