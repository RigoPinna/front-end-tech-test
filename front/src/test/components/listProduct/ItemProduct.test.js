import React from 'react'
import { mount } from 'enzyme'

import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store'

import '@testing-library/jest-dom'
import { Router } from "react-router"
import { ItemProduct } from '../../../components/listProducts/ItemProduct'

const middlewares = [ thunk]
const mockStore = configureStore(middlewares)
const productsReducer = [ 
        {
            "_id": "1",
            "name": "Airpods Wireless Bluetooth Headphones",
            "image": "/images/airpods.jpg",
            "description": "Bluetooth technology lets you connect it with compatible devices wirelessly High-quality AAC audio offers immersive listening experience Built-in microphone allows you to take calls while working",
            "brand": "Apple",
            "category": "Electronics",
            "price": 89.99,
            "countInStock": 10,
            "rating": 4.5,
            "numReviews": 12
        },
        {
            "_id": "6",
            "name": "Amazon Echo Dot 3rd Generation",
            "image": "/images/alexa.jpg",
            "description": "Meet Echo Dot - Our most popular smart speaker with a fabric design. It is our most compact smart speaker that fits perfectly into small space",
            "brand": "Amazon",
            "category": "Electronics",
            "price": 29.99,
            "countInStock": 0,
            "rating": 4,
            "numReviews": 12
        }
]
const cartReducer = []

let store = mockStore({cartReducer})
store.dispatch = jest.fn();
const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
 
  }

const wrapper = mount(
    <Provider store={store}>
        <Router history={historyMock}>
            <ItemProduct {...productsReducer[0]}/> 
        </Router>

    </Provider>
)
describe('Pruebas en el componente <ItemProduct />',() => {

    
    test('Debe de mostrarse correctamente',()=>{
        expect( wrapper).toMatchSnapshot()
        expect(wrapper.find('.__card').exists()).toBe(true)
    })

    test('Debe mostrar el nombre correctamente y el boton "Add to item cart" debe estar activo',()=>{
        expect(wrapper.find('h4').text()).toBe(productsReducer[0].name)
        const buttonAddtoItem = wrapper.find('button').prop('disabled').valueOf()

        expect( buttonAddtoItem ).toBe( false )
    })

    test('Debe deshabilitar el boton "Add to item cart" si el stock es menor que 1',() => {
        const wrapper = mount(
            <Provider store={store}>
                <Router history={historyMock}>
                    <ItemProduct {...productsReducer[1]} />
                </Router>
            </Provider>
        )
        expect( wrapper).toMatchSnapshot()
        const buttonAddtoItem = wrapper.find('button').prop('disabled').valueOf()

        expect( buttonAddtoItem ).toBe( true )
    })
    test('Debe de navegar a la pÃ¡gina del producto al dar clic', () => {
        const item = wrapper.find('.__card');
        item.prop('onClick')();
        expect( historyMock.push ).toHaveBeenCalledTimes(1)
    })
    afterEach(() => {
        store = mockStore({ cartReducer });
        jest.clearAllMocks();
    })
    
})
