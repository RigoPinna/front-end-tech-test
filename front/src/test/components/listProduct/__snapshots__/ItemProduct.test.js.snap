// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <ItemProduct /> Debe de mostrarse correctamente 1`] = `
<div
  className="__card animate__animated animate__fadeIn"
  onClick={[Function]}
>
  <span>
    #
    4.5
  </span>
  <img
    alt="Airpods Wireless Bluetooth Headphones"
    src="http://localhost:5000/images/airpods.jpg"
  />
  <div
    className="__card_body"
  >
    <h4>
      Airpods Wireless Bluetooth Headphones
    </h4>
    <p>
      Apple - Electronics
    </p>
    <p
      className="__card_body_description"
    >
      Bluetooth technology lets you connect it with compatible devices wirelessly High-quality AAC audio offers immersive listening experience Built-in microphone allows you to take calls while working
    </p>
    <div
      className="__wrapper_reviews"
    >
      <fieldset
        className="rating rating--static"
      >
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={1}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={1.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={2}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={2.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={3}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={3.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={4}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={4.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={true}
            name="star"
            type="radio"
            value={5}
          />
          <span />
        </label>
      </fieldset>
      <p>
        (
        12
         Reviwers)
      </p>
    </div>
  </div>
  <div
    className="__card_footer"
  >
    <strong>
      $
      89.99
    </strong>
    <button
      disabled={false}
      onClick={[Function]}
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M11 11V5h2v6h6v2h-6v6h-2v-6H5v-2z"
        />
      </svg>
      Add to item cart
    </button>
  </div>
</div>
`;

exports[`Pruebas en el componente <ItemProduct /> Debe deshabilitar el boton "Add to item cart" si el stock es menor que 1 1`] = `
<div
  className="__card animate__animated animate__fadeIn"
  onClick={[Function]}
>
  <span>
    #
    4
  </span>
  <img
    alt="Amazon Echo Dot 3rd Generation"
    src="http://localhost:5000/images/alexa.jpg"
  />
  <div
    className="__card_body"
  >
    <h4>
      Amazon Echo Dot 3rd Generation
    </h4>
    <p>
      Amazon - Electronics
    </p>
    <p
      className="__card_body_description"
    >
      Meet Echo Dot - Our most popular smart speaker with a fabric design. It is our most compact smart speaker that fits perfectly into small space
    </p>
    <div
      className="__wrapper_reviews"
    >
      <fieldset
        className="rating rating--static"
      >
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={1}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={1.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={2}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={2.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={3}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={3.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={false}
            name="star"
            type="radio"
            value={4}
          />
          <span />
        </label>
        <label>
          <input
            disabled={true}
            name="star"
            type="radio"
            value={4.5}
          />
          <span />
        </label>
        <label>
          <input
            disabled={true}
            name="star"
            type="radio"
            value={5}
          />
          <span />
        </label>
      </fieldset>
      <p>
        (
        12
         Reviwers)
      </p>
    </div>
  </div>
  <div
    className="__card_footer"
  >
    <strong>
      $
      29.99
    </strong>
    <button
      disabled={true}
      onClick={[Function]}
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2c5.523 0 10 4.477 10 10 0 .727-.077 1.435-.225 2.118l-1.782-1.783a8 8 0 10-4.375 6.801 3.997 3.997 0 001.555 1.423A9.956 9.956 0 0112 22C6.477 22 2 17.523 2 12S6.477 2 12 2zm7 12.172l1.414 1.414a2 2 0 11-2.93.11l.102-.11L19 14.172zM12 15c1.466 0 2.785.631 3.7 1.637l-.945.86C13.965 17.182 13.018 17 12 17c-1.018 0-1.965.183-2.755.496l-.945-.86A4.987 4.987 0 0112 15zm-3.5-5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zm7 0a1.5 1.5 0 110 3 1.5 1.5 0 010-3z"
        />
      </svg>
      Out of stock
    </button>
  </div>
</div>
`;
